src/uglify/index.js

```diff
@@ -0,0 +1,83 @@
+import os from 'os';
+import findCacheDir from 'find-cache-dir';
+import workerFarm from 'worker-farm';
+import { get, put } from './cache';
+import { encode } from './serialization';
+import versions from './versions';
+import worker from './worker';
+
+let workerFile = require.resolve('./worker');
+
+try {
+  // run test
+  workerFile = require.resolve('../../dist/uglify/worker');
+} catch (e) { } // eslint-disable-line no-empty
+
+class Uglify {
+  constructor({
+    cache = findCacheDir({ name: 'uglifyjs-webpack-plugin' }),
+    workers = os.cpus().length - 1,
+  }) {
+    this.cache = cache;
+    this.workers = workers;
+  }
+
+  worker(options, callback) {
+    if (this.workers > 0) {
+      this.workerFarm = workerFarm({
+        maxConcurrentWorkers: this.workers,
+      }, workerFile);
+      this.worker = this.workerFarm;
+    } else {
+      this.worker = worker;
+    }
+
+    this.worker(options, callback);
+  }
+
+  exit() {
+    if (this.workerFarm) {
+      workerFarm.end(this.workerFarm);
+    }
+  }
+
+  runTasks(tasks, callback) {
+    let toRun = tasks.length;
+    const results = [];
+    const step = (index, data) => {
+      toRun -= 1;
+      results[index] = data;
+      if (!toRun) {
+        callback(null, results);
+      }
+    };
+    if (!tasks.length) {
+      process.nextTick(() => {
+        callback(null, results);
+      });
+      return;
+    }
+    tasks.forEach((task, index) => {
+      const cacheIdentifier = `${versions.uglify}|${versions.plugin}|${task.input}`;
+      const enqueue = () => {
+        this.worker(JSON.stringify(task, encode), (errors, data) => {
```

Comment:
You shouldn't have to Stringify the task if it's not parallel, but I guess this is fine?